File: sdks/go/verify.go
package authowl

import (
	"context"
	"crypto/ecdsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"math/big"
	"regexp"
	"strings"
	"time"
)

var (
	segmentPattern   = regexp.MustCompile(`^[A-Za-z0-9_-]+$`)
	signaturePattern = regexp.MustCompile(`^[A-Za-z0-9_-]{86}$`)
)

// DefaultClockTolerance is the skew allowed on exp/nbf when none is configured.
const DefaultClockTolerance = 60 * time.Second

// MaxClockTolerance caps the configurable skew. A tolerance larger than this
// keeps revoked tokens alive too long to be called authorization.
const MaxClockTolerance = 300 * time.Second

// VerifiedToken is the result of a successful verification.
type VerifiedToken struct {
	// Subject is the signed-in user id, or "" when the token carries no `sub`.
	Subject string
	// Membership is the active-org membership, or nil when the token carries none.
	Membership *Membership
	// Claims is the full verified claim set, for callers reading extra claims.
	Claims map[string]any
}

// KeySource resolves the verification key named by a token's `kid`.
type KeySource interface {
	// ResolveKey returns the key for kid, or the first published key when kid is
	// empty. It must return a *VerificationError with ErrJWKSKeyNotFound when no
	// key matches.
	ResolveKey(ctx context.Context, kid string) (*JWK, error)
}

// Verifier performs stateless verification of AuthOwl project JWTs.
//
// This is the REAL server-side authorization primitive. It verifies the ES256
// signature against the project's published JWKS and checks issuer, audience,
// and expiry BEFORE reading any claim, so no permission is ever granted off an
// unverified claim.
type Verifier struct {
	// Issuer is the expected `iss`: the project's AuthOwl auth base URL.
	Issuer string
	// Audience is the expected `aud`: the project id.
	Audience string
	// Keys resolves verification keys. Use NewRemoteKeySource in production.
	Keys KeySource
	// ClockTolerance bounds exp/nbf skew, from 0 through MaxClockTolerance.
	//
	// A POINTER so "unset" and "explicitly zero" stay distinct: nil means
	// DefaultClockTolerance, while ClockSkew(0) demands an exact match. A plain
	// duration would let Go's zero value turn the strictest setting into the
	// default one.
	ClockTolerance *time.Duration
	// Now is injectable for deterministic tests. Nil means time.Now.
	Now func() time.Time
}

// ClockSkew returns a pointer for Verifier.ClockTolerance.
func ClockSkew(d time.Duration) *time.Duration { return &d }

func (v *Verifier) now() time.Time {
	if v.Now != nil {
		return v.Now()
	}
	return time.Now()
}

func (v *Verifier) tolerance() time.Duration {
	if v.ClockTolerance == nil {
		return DefaultClockTolerance
	}
	return *v.ClockTolerance
}

func decodeSegment(segment string) (map[string]any, error) {
	malformed := verr(ErrTokenMalformed, "malformed JWT segment")
	if !segmentPattern.MatchString(segment) {
		return nil, malformed
	}
	raw, err := base64.RawURLEncoding.DecodeString(segment)
	if err != nil {
		return nil, malformed
	}
	var claims map[string]any
	if err := json.Unmarshal(raw, &claims); err != nil || claims == nil {
		return nil, malformed
	}
	return claims, nil
}

func audienceMatches(aud any, expected string) bool {
	switch value := aud.(type) {
	case string:
		return value == expected
	case []any:
		for _, entry := range value {
			if text, ok := entry.(string); ok && text == expected {
				return true
			}
		}
	}
	return false
}

func stringSlice(value any) ([]string, bool) {
	entries, ok := value.([]any)
	if !ok {
		return nil, false
	}
	// Non-string entries are DROPPED rather than failing the whole claim, so one
	// malformed entry cannot deny an otherwise valid membership.
	result := make([]string, 0, len(entries))
	for _, entry := range entries {
		if text, ok := entry.(string); ok {
			result = append(result, text)
		}
	}
	return result, true
}

func readMembership(claims map[string]any) *Membership {
	raw, ok := claims["membership"].(map[string]any)
	if !ok {
		return nil
	}
	role, _ := raw["role"].(string)
	permissions, _ := stringSlice(raw["permissions"])
	if permissions == nil {
		permissions = []string{}
	}
	// Absent stays nil rather than becoming an empty slice, so Has({TeamID}) can
	// never be satisfied by a claim that never mentioned teams.
	teams, hasTeams := stringSlice(raw["teams"])
	if !hasTeams {
		teams = nil
	}
	if role == "" && len(permissions) == 0 && len(teams) == 0 {
		return nil
	}
	return &Membership{Role: role, Permissions: permissions, Teams: teams}
}

// Verify checks a project JWT and returns its subject, membership, and claims.
//
// Checks run in a deliberate order - structure, algorithm, key, signature, then
// claims - so a token with a bad signature always reports as a signature
// failure even when its claims are also invalid.
func (v *Verifier) Verify(ctx context.Context, token string) (*VerifiedToken, error) {
	if v.Issuer == "" || v.Audience == "" || v.Keys == nil {
		return nil, verr(ErrTokenConfigInvalid, "verifier requires Issuer, Audience, and Keys")
	}
	if v.ClockTolerance != nil && (*v.ClockTolerance < 0 || *v.ClockTolerance > MaxClockTolerance) {
		return nil, verr(ErrTokenConfigInvalid, "ClockTolerance must be between 0 and 300 seconds")
	}
	if token == "" {
		return nil, verr(ErrTokenMalformed, "a token string is required")
	}
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, verr(ErrTokenMalformed, "malformed JWT")
	}
	headerSegment, payloadSegment, signatureSegment := parts[0], parts[1], parts[2]

	header, err := decodeSegment(headerSegment)
	if err != nil {
		return nil, err
	}
	// The algorithm is pinned BEFORE key resolution, which is what defeats the
	// `alg: none` and HS256-confusion families outright.
	if alg, _ := header["alg"].(string); alg != "ES256" {
		return nil, verr(ErrTokenAlgorithmUnsupported, "unsupported JWT algorithm")
	}
	kid, _ := header["kid"].(string)

	key, err := v.Keys.ResolveKey(ctx, kid)
	if err != nil {
		return nil, err
	}
	if key == nil || key.public == nil {
		return nil, verr(ErrJWKSKeyNotFound, "no matching JWKS key for the token kid")
	}

	if !signaturePattern.MatchString(signatureSegment) {
		return nil, verr(ErrTokenMalformed, "malformed JWT signature")
	}
	signature, decodeErr := base64.RawURLEncoding.DecodeString(signatureSegment)
	if decodeErr != nil || len(signature) != 64 {
		return nil, verr(ErrTokenMalformed, "malformed JWT signature")
	}

	digest := sha256.Sum256([]byte(headerSegment + "." + payloadSegment))
	r := new(big.Int).SetBytes(signature[:32])
	s := new(big.Int).SetBytes(signature[32:])
	if !ecdsa.Verify(key.public, digest[:], r, s) {
		return nil, verr(ErrTokenSignatureInvalid, "invalid token signature")
	}

	claims, err := decodeSegment(payloadSegment)
	if err != nil {
		return nil, err
	}

	now := v.now().Unix()
	tolerance := int64(v.tolerance() / time.Second)

	// `exp` is REQUIRED, not skip-if-absent: a token with no expiry would never
	// fail closed on its own.
	exp, ok := claims["exp"].(float64)
	if !ok {
		return nil, verr(ErrTokenClaimInvalid, "token is missing a valid exp claim")
	}
	if int64(exp)+tolerance < now {
		return nil, verr(ErrTokenClaimInvalid, "token has expired")
	}
	if nbf, present := claims["nbf"].(float64); present && int64(nbf)-tolerance > now {
		return nil, verr(ErrTokenClaimInvalid, "token is not yet valid")
	}
	// `iss` is REQUIRED and must match exactly - including the trailing slash.
	if iss, _ := claims["iss"].(string); iss != v.Issuer {
		return nil, verr(ErrTokenClaimInvalid, "token issuer missing or mismatched")
	}
	if !audienceMatches(claims["aud"], v.Audience) {
		return nil, verr(ErrTokenClaimInvalid, "token audience mismatch")
	}

	subject, _ := claims["sub"].(string)
	return &VerifiedToken{
		Subject:    subject,
		Membership: readMembership(claims),
		Claims:     claims,
	}, nil
}

// Has verifies the token and evaluates the query against its membership.
//
// Fails CLOSED: an invalid, tampered, expired, or wrong-audience token returns
// false rather than an error, so a caller that ignores errors still denies.
// Configuration mistakes are the exception - those return an error, because a
// misconfigured backend silently denying every request is far worse to debug.
func (v *Verifier) Has(ctx context.Context, token string, query Query) (bool, error) {
	if v.Issuer == "" || v.Audience == "" || v.Keys == nil {
		return false, verr(ErrTokenConfigInvalid, "verifier requires Issuer, Audience, and Keys")
	}
	verified, err := v.Verify(ctx, token)
	if err != nil {
		if CodeOf(err) == ErrTokenConfigInvalid {
			return false, err
		}
		return false, nil
	}
	return verified.Membership.Has(query), nil
}

// HasPermission verifies the token and reports whether it grants permission.
// Fails closed exactly like Has.
func (v *Verifier) HasPermission(ctx context.Context, token, permission string) (bool, error) {
	return v.Has(ctx, token, Query{Permission: Match(permission)})
}

File: sdks/rust/src/verify.rs
//! Stateless verification of AuthOwl project JWTs.
//!
//! This is the REAL server-side authorization primitive. It verifies the ES256
//! signature against the project's published JWKS and checks issuer, audience,
//! and expiry BEFORE reading any claim, so no permission is ever granted off an
//! unverified claim.

use std::time::{SystemTime, UNIX_EPOCH};

use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine as _;
use p256::ecdsa::signature::Verifier as _;
use p256::ecdsa::Signature;
use serde_json::{Map, Value};

use crate::error::{ErrorCode, VerificationError};
use crate::jwks::{is_base64url_byte, KeySource};
use crate::membership::{Membership, Query};

/// Skew allowed on `exp`/`nbf` when none is configured.
pub const DEFAULT_CLOCK_TOLERANCE_SECONDS: u64 = 60;
/// A tolerance beyond this keeps revoked tokens alive too long to be called
/// authorization, so it is refused as a configuration error.
pub const MAX_CLOCK_TOLERANCE_SECONDS: u64 = 300;

/// The declared purpose of a project JWT.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum TokenUse {
    Session,
    Template,
    Access,
    Id,
}

/// The result of a successful verification.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct VerifiedToken {
    /// The signed-in user id, or `None` when the token carries no `sub`.
    pub subject: Option<String>,
    /// The active-org membership, or `None` when the token carries none.
    pub membership: Option<Membership>,
    /// The full verified claim set, for callers reading additional claims.
    pub claims: Map<String, Value>,
}

/// Verifies AuthOwl project JWTs against a project's published keys.
pub struct Verifier<K: KeySource> {
    issuer: String,
    audience: String,
    keys: K,
    clock_tolerance_seconds: u64,
    token_use: Option<TokenUse>,
    require_token_use: bool,
}

// Hand-written so a `KeySource` need not be `Debug` itself, and so the key
// material behind it never reaches a log line.
impl<K: KeySource> std::fmt::Debug for Verifier<K> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Verifier")
            .field("issuer", &self.issuer)
            .field("audience", &self.audience)
            .field("clock_tolerance_seconds", &self.clock_tolerance_seconds)
            .field("token_use", &self.token_use)
            .field("require_token_use", &self.require_token_use)
            .finish_non_exhaustive()
    }
}

impl<K: KeySource> Verifier<K> {
    /// Build a verifier. Fails when the configuration itself is unusable.
    pub fn new(
        issuer: impl Into<String>,
        audience: impl Into<String>,
        keys: K,
    ) -> Result<Self, VerificationError> {
        Self::with_clock_tolerance(issuer, audience, keys, DEFAULT_CLOCK_TOLERANCE_SECONDS)
    }

    pub fn with_clock_tolerance(
        issuer: impl Into<String>,
        audience: impl Into<String>,
        keys: K,
        clock_tolerance_seconds: u64,
    ) -> Result<Self, VerificationError> {
        let issuer = issuer.into();
        let audience = audience.into();
        if issuer.is_empty() || audience.is_empty() {
            return Err(VerificationError::new(
                ErrorCode::TokenConfigInvalid,
                "verifier requires an issuer and an audience",
            ));
        }
        if clock_tolerance_seconds > MAX_CLOCK_TOLERANCE_SECONDS {
            return Err(VerificationError::new(
                ErrorCode::TokenConfigInvalid,
                "clock tolerance must be from 0 through 300 seconds",
            ));
        }
        Ok(Self {
            issuer,
            audience,
            keys,
            clock_tolerance_seconds,
            token_use: None,
            require_token_use: false,
        })
    }

    /// Narrow verification to one declared token purpose.
    pub fn with_token_use(mut self, token_use: TokenUse) -> Self {
        self.token_use = Some(token_use);
        self
    }

    /// Reject legacy tokens that do not carry a `token_use` claim.
    pub fn with_required_token_use(mut self, required: bool) -> Self {
        self.require_token_use = required;
        self
    }

    /// Verify a token against the system clock.
    pub fn verify(&self, token: &str) -> Result<VerifiedToken, VerificationError> {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|elapsed| elapsed.as_secs() as i64)
            .unwrap_or(0);
        self.verify_at(token, now)
    }

    /// Verify a token against an explicit Unix timestamp.
    ///
    /// Taking the clock as a parameter rather than injecting a closure keeps the
    /// verifier free of interior mutability and makes tests deterministic.
    ///
    /// Checks run in a deliberate order - structure, algorithm, key, signature,
    /// then claims - so a token with a bad signature always reports as a
    /// signature failure even when its claims are also invalid.
    pub fn verify_at(&self, token: &str, now: i64) -> Result<VerifiedToken, VerificationError> {
        self.verify_at_for_token_use(token, now, self.token_use)
    }

    fn verify_at_for_token_use(
        &self,
        token: &str,
        now: i64,
        accepted_token_use: Option<TokenUse>,
    ) -> Result<VerifiedToken, VerificationError> {
        if token.is_empty() {
            return Err(VerificationError::new(
                ErrorCode::TokenMalformed,
                "a token string is required",
            ));
        }
        let parts: Vec<&str> = token.split('.').collect();
        if parts.len() != 3 {
            return Err(VerificationError::new(
                ErrorCode::TokenMalformed,
                "malformed JWT",
            ));
        }
        let (header_segment, payload_segment, signature_segment) = (parts[0], parts[1], parts[2]);

        let header = decode_segment(header_segment)?;
        // The algorithm is pinned BEFORE key resolution, which is what defeats
        // the `alg: none` and HS256-confusion families outright.
        if header.get("alg").and_then(Value::as_str) != Some("ES256") {
            return Err(VerificationError::new(
                ErrorCode::TokenAlgorithmUnsupported,
                "unsupported JWT algorithm",
            ));
        }
        let kid = header.get("kid").and_then(Value::as_str);
        let key = self.keys.resolve_key(kid)?;

        let malformed_signature =
            || VerificationError::new(ErrorCode::TokenMalformed, "malformed JWT signature");
        if signature_segment.len() != 86 || !signature_segment.bytes().all(is_base64url_byte) {
            return Err(malformed_signature());
        }
        let raw_signature = URL_SAFE_NO_PAD
            .decode(signature_segment)
            .map_err(|_| malformed_signature())?;
        if raw_signature.len() != 64 {
            return Err(malformed_signature());
        }
        let signature = Signature::from_slice(&raw_signature).map_err(|_| malformed_signature())?;

        let signing_input = format!("{header_segment}.{payload_segment}");
        key.verifying_key()
            .verify(signing_input.as_bytes(), &signature)
            .map_err(|_| {
                VerificationError::new(ErrorCode::TokenSignatureInvalid, "invalid token signature")
            })?;

        let claims = decode_segment(payload_segment)?;
        let tolerance = self.clock_tolerance_seconds as i64;

        // `exp` is REQUIRED, not skip-if-absent: a token with no expiry would
        // never fail closed on its own.
        let exp = numeric_claim(claims.get("exp")).ok_or_else(|| {
            VerificationError::new(
                ErrorCode::TokenClaimInvalid,
                "token is missing a valid exp claim",
            )
        })?;
        if exp + tolerance < now {
            return Err(VerificationError::new(
                ErrorCode::TokenClaimInvalid,
                "token has expired",
            ));
        }
        if let Some(nbf) = numeric_claim(claims.get("nbf")) {
            if nbf - tolerance > now {
                return Err(VerificationError::new(
                    ErrorCode::TokenClaimInvalid,
                    "token is not yet valid",
                ));
            }
        }
        // `iss` is REQUIRED and must match exactly - including any trailing slash.
        if claims.get("iss").and_then(Value::as_str) != Some(self.issuer.as_str()) {
            return Err(VerificationError::new(
                ErrorCode::TokenClaimInvalid,
                "token issuer missing or mismatched",
            ));
        }
        if !audience_matches(claims.get("aud"), &self.audience) {
            return Err(VerificationError::new(
                ErrorCode::TokenClaimInvalid,
                "token audience mismatch",
            ));
        }
        validate_token_use(
            header.get("typ"),
            claims.get("token_use"),
            accepted_token_use,
            self.require_token_use,
        )?;

        Ok(VerifiedToken {
            subject: claims.get("sub").and_then(Value::as_str).map(str::to_owned),
            membership: Membership::from_claim(claims.get("membership")),
            claims,
        })
    }

    /// Verify the token, then evaluate the query against its membership.
    ///
    /// Fails CLOSED: an invalid, tampered, expired, or wrong-audience token
    /// returns `false` rather than an error. Configuration mistakes are rejected
    /// in [`Verifier::new`], so a misconfigured verifier can never reach this
    /// silent-denial path.
    pub fn has(&self, token: &str, query: &Query<'_>) -> bool {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|elapsed| elapsed.as_secs() as i64)
            .unwrap_or(0);
        self.has_at(token, query, now)
    }

    pub fn has_at(&self, token: &str, query: &Query<'_>, now: i64) -> bool {
        match self.verify_at_for_token_use(token, now, Some(TokenUse::Session)) {
            Ok(verified) => verified.membership.is_some_and(|m| m.has(query)),
            Err(_) => false,
        }
    }

    /// Verify the token and report whether it grants `permission`. Fails closed.
    pub fn has_permission(&self, token: &str, permission: &str) -> bool {
        self.has(token, &Query::permission(permission))
    }
}

fn validate_token_use(
    typ: Option<&Value>,
    raw: Option<&Value>,
    accepted: Option<TokenUse>,
    required: bool,
) -> Result<(), VerificationError> {
    let unsupported = || {
        VerificationError::new(
            ErrorCode::TokenUseUnsupported,
            "token purpose is missing or unsupported",
        )
    };
    let Some(raw) = raw else {
        if required || typ.and_then(Value::as_str) != Some("JWT") {
            return Err(unsupported());
        }
        return Ok(());
    };
    let Some(value) = raw.as_str() else {
        return Err(unsupported());
    };
    let use_kind = match value {
        "session" => TokenUse::Session,
        "template" => TokenUse::Template,
        "access" => TokenUse::Access,
        "id" => TokenUse::Id,
        _ => return Err(unsupported()),
    };
    if accepted.is_none() && use_kind == TokenUse::Id
        || accepted.is_some_and(|expected| expected != use_kind)
    {
        return Err(unsupported());
    }
    let expected_typ = if use_kind == TokenUse::Access {
        "at+jwt"
    } else {
        "JWT"
    };
    if typ.and_then(Value::as_str) != Some(expected_typ) {
        return Err(unsupported());
    }
    Ok(())
}

fn decode_segment(segment: &str) -> Result<Map<String, Value>, VerificationError> {
    let malformed = || VerificationError::new(ErrorCode::TokenMalformed, "malformed JWT segment");
    if segment.is_empty() || !segment.bytes().all(is_base64url_byte) {
        return Err(malformed());
    }
    let raw = URL_SAFE_NO_PAD.decode(segment).map_err(|_| malformed())?;
    let value: Value = serde_json::from_slice(&raw).map_err(|_| malformed())?;
    match value {
        Value::Object(map) => Ok(map),
        _ => Err(malformed()),
    }
}

/// Read a numeric claim.
///
/// `serde_json` keeps booleans out of `as_f64`, so `"exp": true` correctly
/// reports as absent rather than sliding through as the number 1.
fn numeric_claim(value: Option<&Value>) -> Option<i64> {
    value?.as_f64().map(|number| number as i64)
}

fn audience_matches(aud: Option<&Value>, expected: &str) -> bool {
    match aud {
        Some(Value::String(value)) => value == expected,
        Some(Value::Array(entries)) => entries.iter().any(|entry| entry.as_str() == Some(expected)),
        _ => false,
    }
}
